File: scripts/analyze_benchmark.py
#!/usr/bin/env python3
"""
Analyze RefRover MAG benchmark results — the Tier-1 proxy vs Tier-2 yield question.

Reads ``benchmark_results.tsv`` (from ``refrover aggregate-results``) and answers
the open research question in CLAUDE.md / PLAN.md §8:

    Does the cheap Tier-1 proxy (frac_variance) rank selection configurations the
    same way real MAG yield (Tier-2) does — so we can pick a selector *without*
    paying for alignment + MetaBAT2 + CheckM2?

Why this script exists alongside ``analyze_benchmark.R``:
  * The R script makes the plots but (a) does ``k <- as.integer(k)``, which turns
    every ``k="adaptive"`` row into NA, and (b) reports a single *pooled* Spearman
    that is the wrong statistic (see below). This script fixes both and prints the
    correct decomposition. No R/tidyverse install needed — pandas/numpy/scipy only.

The statistical point
---------------------
``frac_variance`` is the fraction of *that matrix's own* total cross-sample
variance a selection explains. The containment matrix and the Jaccard matrix have
different total variances, so frac_variance is **not comparable across matrices** —
a pooled correlation over both matrices is driven by the between-matrix mean offset
(a Simpson's-paradox confound), not by whether the proxy tracks yield. Likewise the
focal assembly is a massive blocking factor (some focals yield ~50 MAGs, others
~30), so any correlation must condition on focal too.

So we report, for each proxy × yield metric:
  1. POOLED Spearman            — the confounded number, shown only as a warning.
  2. WITHIN-MATRIX Spearman     — removes the cross-matrix scale confound.
  3. WITHIN-(matrix, focal)     — the fair test: does the proxy pick the right
                                  rule/k when matrix & focal are held fixed?
  4. TOP-1 HIT RATE             — per (matrix, focal), does argmax(proxy) select
                                  the same config as argmax(yield)? The most
                                  decision-relevant summary: "if I trust the proxy
                                  to pick the config, how often do I get the real
                                  winner?"

Also emitted: the mean-yield leaderboard, a Jaccard-vs-containment cross-matrix
agreement summary, and an adaptive-k vs fixed-k comparison.

Usage
-----
    python scripts/analyze_benchmark.py \
        --results work/benchmark3/benchmark_results.tsv \
        --outdir  work/benchmark3/analysis/

Outputs (TSV, plus a printed summary):
    leaderboard.tsv            mean/median yield per (matrix, rule, k_label)
    proxy_correlations.tsv     pooled / within-matrix / within-focal Spearman
    top1_hit_rate.tsv          proxy-picks-real-winner rate per matrix
    cross_matrix.tsv           jaccard vs containment yield agreement per rule
    adaptive_vs_fixed.tsv      adaptive-k vs fixed-k yield per matrix, rule
"""
from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np
import pandas as pd
from scipy.stats import spearmanr

PROXY_COLS = ["frac_variance", "effective_rank", "tiered_axis_count"]
MIN_N_FOR_R = 4  # need at least this many points for a meaningful Spearman


# ── load & normalize ──────────────────────────────────────────────────────────

CELL_KEY = ["matrix", "rule", "k", "focal"]


def load(path: Path) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Split aggregated results into (successful cells, error cells).

    Defensively drops stale error rows: a cell that failed once then succeeded on
    a --resume retry can appear as both (older aggregate_results globbed error.json
    independently of result.json). A success for a cell is authoritative — an error
    row for the same (matrix, rule, k, focal) is discarded so the failure report
    isn't inflated and the cell isn't double-counted.
    """
    raw = pd.read_csv(path, sep="\t")
    if "error" not in raw.columns:
        return raw.copy(), raw.iloc[0:0].copy()

    ok = raw[raw["error"].isna()].copy()
    err = raw[raw["error"].notna()].copy()
    if set(CELL_KEY).issubset(raw.columns) and not err.empty:
        succeeded = set(map(tuple, ok[CELL_KEY].astype(str).itertuples(index=False)))
        keep = ~err[CELL_KEY].astype(str).apply(tuple, axis=1).isin(succeeded)
        err = err[keep].copy()
    return ok, err


def add_k_columns(df: pd.DataFrame) -> pd.DataFrame:
    """Resolve ``k`` so adaptive rows survive.

    ``k_label`` is the original string ("10" or "adaptive") — the axis you group
    by. ``k_eff`` is the integer actually used: ``k_actual`` for adaptive rows,
    else ``int(k)``. The R script's ``as.integer(k)`` silently NA's "adaptive";
    this keeps both views.
    """
    df = df.copy()
    df["k_label"] = df["k"].astype(str)
    k_actual = pd.to_numeric(df.get("k_actual"), errors="coerce")
    k_int = pd.to_numeric(df["k"], errors="coerce")
    df["k_eff"] = k_actual.where(df["k_label"] == "adaptive", k_int)
    return df


def pick_yield_metrics(df: pd.DataFrame) -> list[str]:
    """Yield metrics to analyze, primary first. sum_qs preferred over weighted_mags."""
    return [c for c in ("sum_qs", "weighted_mags") if c in df.columns]


# ── analyses ──────────────────────────────────────────────────────────────────

def leaderboard(df: pd.DataFrame, yvar: str) -> pd.DataFrame:
    out = (
        df.groupby(["matrix", "rule", "k_label"], dropna=False)[yvar]
        .agg(mean="mean", median="median", std="std", n="count")
        .reset_index()
        .sort_values("mean", ascending=False, ignore_index=True)
    )
    return out


def _grouped_spearman(df: pd.DataFrame, group_cols: list[str],
                      proxy: str, yvar: str) -> pd.DataFrame:
    """Spearman(proxy, yvar) within each group; one row per group meeting MIN_N."""
    rows = []
    for key, g in df.groupby(group_cols, dropna=False):
        x = g[proxy].to_numpy(dtype=float)
        y = g[yvar].to_numpy(dtype=float)
        ok = np.isfinite(x) & np.isfinite(y)
        if ok.sum() < MIN_N_FOR_R or np.ptp(x[ok]) == 0 or np.ptp(y[ok]) == 0:
            continue
        r, p = spearmanr(x[ok], y[ok])
        rec = dict(zip(group_cols, key if isinstance(key, tuple) else (key,)))
        rec.update(rho=r, p=p, n=int(ok.sum()))
        rows.append(rec)
    return pd.DataFrame(rows)


def proxy_correlations(df: pd.DataFrame, yvar: str) -> pd.DataFrame:
    """Pooled / within-matrix / within-(matrix,focal) Spearman for each proxy.

    The within-focal row is an n-weighted mean of per-(matrix,focal) rho — the
    fair test. Pooled is included but flagged: it is confounded by the cross-matrix
    frac_variance scale offset and by between-focal yield differences.
    """
    proxies = [c for c in PROXY_COLS if c in df.columns]
    rows = []
    for proxy in proxies:
        x = df[proxy].to_numpy(dtype=float)
        y = df[yvar].to_numpy(dtype=float)
        ok = np.isfinite(x) & np.isfinite(y)
        if ok.sum() >= MIN_N_FOR_R:
            r, p = spearmanr(x[ok], y[ok])
            rows.append(dict(proxy=proxy, scope="pooled (CONFOUNDED)",
                             rho=r, p=p, n=int(ok.sum()), n_groups=1))

        per_matrix = _grouped_spearman(df, ["matrix"], proxy, yvar)
        for _, r in per_matrix.iterrows():
            rows.append(dict(proxy=proxy, scope=f"within-matrix[{r['matrix']}]",
                             rho=r["rho"], p=r["p"], n=int(r["n"]), n_groups=1))

        per_mf = _grouped_spearman(df, ["matrix", "focal"], proxy, yvar)
        if not per_mf.empty:
            w = per_mf["n"]
            rows.append(dict(
                proxy=proxy, scope="within-(matrix,focal) [FAIR TEST, n-wtd mean]",
                rho=float(np.average(per_mf["rho"], weights=w)),
                p=np.nan, n=int(w.sum()), n_groups=int(len(per_mf)),
            ))
    return pd.DataFrame(rows)


def top1_hit_rate(df: pd.DataFrame, yvar: str) -> pd.DataFrame:
    """Per (matrix, focal): does the config with max proxy == config with max yield?

    A config here is a (rule, k_label) combination. Hit rate is the fraction of
    (matrix, focal) blocks where trusting the proxy to pick the config would have
    landed on the real yield winner. Reported per matrix and per proxy.
    """
    proxies = [c for c in PROXY_COLS if c in df.columns]
    rows = []
    for proxy in proxies:
        for matrix, gm in df.groupby("matrix"):
            hits = total = 0
            for _, g in gm.groupby("focal"):
                g = g.dropna(subset=[proxy, yvar])
                if len(g) < 2:
                    continue
                total += 1
                proxy_pick = g.loc[g[proxy].idxmax(), ["rule", "k_label"]]
                yield_win = g.loc[g[yvar].idxmax(), ["rule", "k_label"]]
                if proxy_pick.equals(yield_win):
                    hits += 1
            if total:
                rows.append(dict(proxy=proxy, matrix=matrix,
                                 hit_rate=hits / total, n_focals=total))
    return pd.DataFrame(rows)


def cross_matrix(df: pd.DataFrame, yvar: str) -> pd.DataFrame:
    """Jaccard-vs-containment yield agreement per rule (paired on rule,k_label,focal)."""
    wide = (
        df.pivot_table(index=["rule", "k_label", "focal"], columns="matrix",
                       values=yvar, aggfunc="mean")
        .reset_index()
    )
    if not {"jaccard", "containment"}.issubset(wide.columns):
        return pd.DataFrame()
    rows = []
    for rule, g in wide.groupby("rule"):
        pair = g[["jaccard", "containment"]].dropna()
        if len(pair) < MIN_N_FOR_R:
            continue
        r, p = spearmanr(pair["jaccard"], pair["containment"])
        rows.append(dict(
            rule=rule, rho=r, p=p, n=len(pair),
            mean_jaccard=pair["jaccard"].mean(),
            mean_containment=pair["containment"].mean(),
            containment_minus_jaccard=pair["containment"].mean() - pair["jaccard"].mean(),
        ))
    return pd.DataFrame(rows)


def adaptive_vs_fixed(df: pd.DataFrame, yvar: str) -> pd.DataFrame:
    """Adaptive-k vs each fixed-k yield per (matrix, rule), plus mean k chosen."""
    out = (
        df.groupby(["matrix", "rule", "k_label"], dropna=False)
        .agg(mean_yield=(yvar, "mean"), mean_k_eff=("k_eff", "mean"), n=(yvar, "count"))
        .reset_index()
        .sort_values(["matrix", "rule", "k_label"], ignore_index=True)
    )
    return out


# ── driver ────────────────────────────────────────────────────────────────────

def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--results", required=True, type=Path,
                    help="benchmark_results.tsv from refrover aggregate-results")
    ap.add_argument("--outdir", type=Path, default=Path("benchmark_analysis"))
    args = ap.parse_args()
    args.outdir.mkdir(parents=True, exist_ok=True)

    ok, err = load(args.results)
    ok = add_k_columns(ok)
    metrics = pick_yield_metrics(ok)
    if not metrics:
        raise SystemExit("No yield column (sum_qs / weighted_mags) found in results.")
    yvar = metrics[0]

    print(f"Loaded {len(ok) + len(err)} cells: {len(ok)} ok, {len(err)} errors")
    if len(err):
        cols = [c for c in ["matrix", "rule", "k", "focal", "error"] if c in err.columns]
        print("  Failed cells:")
        print(err[cols].to_string(index=False))
    print(f"Primary yield metric: {yvar}"
          f"{'  (weighted_mags fallback — sum_qs absent)' if yvar != 'sum_qs' else ''}\n")

    lb = leaderboard(ok, yvar)
    corr = proxy_correlations(ok, yvar)
    hits = top1_hit_rate(ok, yvar)
    xmat = cross_matrix(ok, yvar)
    avf = adaptive_vs_fixed(ok, yvar)

    lb.to_csv(args.outdir / "leaderboard.tsv", sep="\t", index=False)
    corr.to_csv(args.outdir / "proxy_correlations.tsv", sep="\t", index=False)
    hits.to_csv(args.outdir / "top1_hit_rate.tsv", sep="\t", index=False)
    xmat.to_csv(args.outdir / "cross_matrix.tsv", sep="\t", index=False)
    avf.to_csv(args.outdir / "adaptive_vs_fixed.tsv", sep="\t", index=False)

    def show(title, frame):
        print(f"── {title} " + "─" * max(0, 60 - len(title)))
        print("(none)" if frame.empty else frame.to_string(index=False))
        print()

    show(f"Leaderboard (top 12 by mean {yvar})", lb.head(12))
    show(f"Proxy → {yvar}: Spearman decomposition", corr)
    show(f"Top-1 hit rate (proxy picks the real {yvar} winner)", hits)
    show(f"Cross-matrix agreement (jaccard vs containment {yvar})", xmat)
    show(f"Adaptive-k vs fixed-k (mean {yvar})", avf)

    print(f"All tables → {args.outdir}/")
    print("\nRead the correlations top-down: trust WITHIN-(matrix,focal) and TOP-1 "
          "HIT RATE.\nA large POOLED rho with weak within-focal rho = the proxy is "
          "only separating\nthe matrices, not ranking configs — do NOT use it to "
          "choose the matrix.")


if __name__ == "__main__":
    main()

File: scripts/analyze_benchmark.R
#!/usr/bin/env Rscript
# Analyze RefRover benchmark results.
#
# Reads benchmark_results.tsv (from refrover aggregate-results) and answers:
#   1. Does selection rule matter?         heatmap + boxplots by rule
#   2. Does k matter?                      yield curve vs k per rule
#   3. Do proxy scores predict MAG yield?  scatter vs primary yield metric + Spearman r
#   4. Jaccard vs containment matrix?      cross-matrix agreement scatter
#   5. Which focal assemblies are hardest? per-focal variability boxplots
#
# Primary yield metric: sum_qs (Σ max(0, completeness − 5×contamination), Olm et al.
# 2017 ISME J) when present; falls back to weighted_mags (2×n_high + 1×n_medium).
#
# Usage:
#   Rscript scripts/analyze_benchmark.R \
#       --results work/benchmark/benchmark_results.tsv \
#       --outdir  work/analysis/
#
# Dependencies: tidyverse, patchwork
#   conda install -n refrover-benchmark -c conda-forge r-base r-tidyverse r-patchwork

suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
})

# ── argument parsing ──────────────────────────────────────────────────────────

args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  out <- list(results = NULL, outdir = "analysis")
  i <- 1
  while (i <= length(args)) {
    if      (args[i] == "--results") { out$results <- args[i + 1]; i <- i + 2 }
    else if (args[i] == "--outdir")  { out$outdir  <- args[i + 1]; i <- i + 2 }
    else { cat("Unknown argument:", args[i], "\n"); i <- i + 1 }
  }
  out
}

opt <- parse_args(args)
if (is.null(opt$results)) {
  cat("Usage: Rscript analyze_benchmark.R --results <tsv> [--outdir <dir>]\n")
  quit(status = 1)
}

dir.create(opt$outdir, recursive = TRUE, showWarnings = FALSE)

# ── colour palette ────────────────────────────────────────────────────────────

rule_colors <- c(
  random     = "#999999",
  maxmin     = "#4477AA",
  css        = "#EE6677",
  kmedoids   = "#228833",
  archetype  = "#CCBB44",
  greedy_var = "#AA3377"
)

# ── load and clean data ───────────────────────────────────────────────────────

cat("Loading", opt$results, "\n")
raw <- read_tsv(opt$results, show_col_types = FALSE)

# Separate successful cells from errors.
if ("error" %in% names(raw)) {
  df  <- filter(raw, is.na(error))
  n_err <- sum(!is.na(raw$error))
  cat(sprintf("Loaded %d cells: %d succeeded, %d errors\n",
              nrow(raw), nrow(df), n_err))
  if (n_err > 0)
    cat("  Error cells excluded — check error.json files for details\n")
} else {
  df <- raw
  cat(sprintf("Loaded %d cells\n", nrow(df)))
}

df <- df |>
  mutate(
    rule   = factor(rule),
    matrix = factor(matrix),
    k      = as.integer(k)
  )

# Determine which yield metric to use as the primary outcome.
# sum_qs  — continuous Quality Score (Olm et al. 2017): preferred, more defensible.
# weighted_mags — binary tier score (2×n_high + 1×n_medium): legacy fallback.
# If both present, plots show sum_qs as primary; weighted_mags available for comparison.
if ("sum_qs" %in% names(df)) {
  yvar  <- "sum_qs"
  ylabel <- "Quality Score sum  (Σ max(0, completeness − 5×contamination))"
  cat("Primary yield metric: sum_qs  (continuous, preferred)\n")
} else {
  yvar  <- "weighted_mags"
  ylabel <- "Weighted MAGs  (2×n_high + 1×n_medium)"
  cat("Primary yield metric: weighted_mags  (sum_qs not found in data)\n")
}

# ── helper: Spearman label ────────────────────────────────────────────────────

spearman_label <- function(x, y) {
  ok <- is.finite(x) & is.finite(y)
  if (sum(ok) < 5) return("r = n/a")
  t     <- cor.test(x[ok], y[ok], method = "spearman", exact = FALSE)
  p_str <- if (t$p.value < 0.001) "p<0.001" else sprintf("p=%.3f", t$p.value)
  sprintf("r = %.2f, %s", t$estimate, p_str)
}

# ── plot 1: heatmap ───────────────────────────────────────────────────────────

cat("Plot 1: heatmap\n")
heat_data <- df |>
  group_by(matrix, rule, k) |>
  summarise(mean_yield = mean(.data[[yvar]], na.rm = TRUE), .groups = "drop")

p1 <- ggplot(heat_data, aes(x = factor(k), y = rule, fill = mean_yield)) +
  geom_tile(colour = "white", linewidth = 0.5) +
  geom_text(aes(label = sprintf("%.1f", mean_yield)), size = 3.2) +
  scale_fill_distiller(palette = "YlOrRd", direction = 1,
                       name = sprintf("mean\n%s", yvar)) +
  facet_wrap(~matrix, labeller = label_both) +
  labs(title = sprintf("Mean %s per (matrix, rule, k)", yvar),
       x = "k  (references per focal assembly)", y = NULL) +
  theme_minimal(base_size = 11) +
  theme(panel.grid = element_blank())

ggsave(file.path(opt$outdir, "01_heatmap.png"), p1,
       width = 5 * n_distinct(df$matrix), height = 4, dpi = 150)

# ── plot 2: k curves ──────────────────────────────────────────────────────────

cat("Plot 2: k curves\n")
p2 <- df |>
  group_by(matrix, rule, k) |>
  summarise(
    mean = mean(.data[[yvar]], na.rm = TRUE),
    sem  = sd(.data[[yvar]],   na.rm = TRUE) / sqrt(n()),
    .groups = "drop"
  ) |>
  ggplot(aes(x = k, y = mean, colour = rule, fill = rule)) +
  geom_ribbon(aes(ymin = mean - sem, ymax = mean + sem),
              alpha = 0.15, colour = NA) +
  geom_line(linewidth = 1) +
  geom_point(size = 2.5) +
  scale_colour_manual(values = rule_colors, aesthetics = c("colour", "fill")) +
  facet_wrap(~matrix, labeller = label_both) +
  labs(title = sprintf("%s vs k  (mean ± SEM over focal assemblies)", yvar),
       x = "k  (references per focal assembly)",
       y = ylabel,
       colour = "rule", fill = "rule") +
  theme_bw(base_size = 11)

ggsave(file.path(opt$outdir, "02_k_curves.png"), p2,
       width = 5 * n_distinct(df$matrix), height = 4, dpi = 150)

# ── plot 3: proxy score correlations ─────────────────────────────────────────

proxy_cols <- intersect(
  c("frac_variance", "effective_rank", "tiered_axis_count"),
  names(df)
)

if (length(proxy_cols) == 0) {
  cat("  No proxy score columns found — skipping plot 3\n")
} else {
  cat("Plot 3: proxy correlations\n")

  proxy_long <- df |>
    select(matrix, rule, k, focal, all_of(c(yvar, proxy_cols))) |>
    pivot_longer(all_of(proxy_cols), names_to = "proxy", values_to = "score")

  spearman_labels <- proxy_long |>
    group_by(matrix, proxy) |>
    summarise(
      label = spearman_label(score, .data[[yvar]]),
      .groups = "drop"
    ) |>
    mutate(x = -Inf, y = Inf, hjust = -0.1, vjust = 1.4)

  p3 <- ggplot(proxy_long,
               aes(x = score, y = .data[[yvar]], colour = rule)) +
    geom_point(alpha = 0.35, size = 1.2) +
    geom_smooth(method = "lm", se = FALSE, aes(group = 1),
                colour = "black", linewidth = 0.7, linetype = "dashed") +
    geom_text(data = spearman_labels,
              aes(x = x, y = y, label = label, hjust = hjust, vjust = vjust),
              colour = "black", size = 2.8, inherit.aes = FALSE) +
    scale_colour_manual(values = rule_colors) +
    facet_grid(proxy ~ matrix,
               scales = "free_x",
               labeller = labeller(matrix = label_both, proxy = label_value)) +
    labs(title = "Proxy scores vs real MAG yield",
         x = "proxy score value", y = ylabel,
         colour = "rule") +
    theme_bw(base_size = 10)

  ggsave(file.path(opt$outdir, "03_proxy_correlations.png"), p3,
         width = 4.5 * n_distinct(df$matrix),
         height = 3.5 * length(proxy_cols), dpi = 150)
}

# ── plot 4: Jaccard vs containment matrix agreement ──────────────────────────

if (n_distinct(df$matrix) < 2) {
  cat("  Only one matrix present — skipping plot 4\n")
} else {
  cat("Plot 4: matrix comparison\n")

  mat_wide <- df |>
    select(rule, k, focal, matrix, all_of(yvar)) |>
    pivot_wider(names_from = matrix, values_from = all_of(yvar)) |>
    drop_na()

  if (!all(c("jaccard", "containment") %in% names(mat_wide))) {
    cat("  Expected matrix names 'jaccard' and 'containment' — skipping plot 4\n")
  } else {
    spearman_mat <- mat_wide |>
      group_by(rule) |>
      summarise(
        label = spearman_label(jaccard, containment),
        .groups = "drop"
      ) |>
      mutate(x = -Inf, y = Inf, hjust = -0.1, vjust = 1.4)

    p4 <- ggplot(mat_wide, aes(x = jaccard, y = containment, colour = rule)) +
      geom_abline(slope = 1, intercept = 0,
                  linetype = "dashed", colour = "grey50", linewidth = 0.7) +
      geom_point(alpha = 0.45, size = 1.8) +
      geom_text(data = spearman_mat,
                aes(x = x, y = y, label = label, hjust = hjust, vjust = vjust),
                colour = "black", size = 3, inherit.aes = FALSE) +
      scale_colour_manual(values = rule_colors) +
      facet_wrap(~rule) +
      labs(title = sprintf("%s: Jaccard vs Containment feature matrix", yvar),
           subtitle = "Each point = one focal assembly × k cell",
           x = sprintf("Jaccard matrix  (%s)", yvar),
           y = sprintf("Containment matrix  (%s)", yvar),
           colour = "rule") +
      theme_bw(base_size = 11)

    ggsave(file.path(opt$outdir, "04_matrix_comparison.png"), p4,
           width = 4 * n_distinct(mat_wide$rule), height = 4.5, dpi = 150)
  }
}

# ── plot 5: per-focal variability ─────────────────────────────────────────────

cat("Plot 5: focal variability\n")
focal_order <- df |>
  group_by(focal) |>
  summarise(med = median(.data[[yvar]], na.rm = TRUE), .groups = "drop") |>
  arrange(desc(med)) |>
  pull(focal)

p5 <- df |>
  mutate(focal = factor(focal, levels = focal_order)) |>
  ggplot(aes(x = focal, y = .data[[yvar]])) +
  geom_boxplot(fill = "#4477AA", alpha = 0.55, outlier.size = 0.8,
               linewidth = 0.4) +
  labs(title = "Per-focal assembly variability  (across all rules × k × matrix)",
       x = "focal assembly  (ordered by median yield)",
       y = ylabel) +
  theme_bw(base_size = 10) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 6))

ggsave(file.path(opt$outdir, "05_focal_variability.png"), p5,
       width = max(8, n_distinct(df$focal) * 0.35), height = 4.5, dpi = 150)

# ── text summary ──────────────────────────────────────────────────────────────

cat(sprintf("\n── Mean %s by (matrix, rule, k) ──────────────────────────\n", yvar))
df |>
  group_by(matrix, rule, k) |>
  summarise(
    mean = round(mean(.data[[yvar]], na.rm = TRUE), 2),
    sd   = round(sd(.data[[yvar]],   na.rm = TRUE), 2),
    n    = n(),
    .groups = "drop"
  ) |>
  arrange(desc(mean)) |>
  print(n = 30)

if (length(proxy_cols) > 0) {
  cat(sprintf("\n── Spearman correlation: proxy score vs %s ──────────\n", yvar))
  df |>
    pivot_longer(all_of(proxy_cols), names_to = "proxy", values_to = "score") |>
    group_by(matrix, proxy) |>
    summarise(
      spearman = spearman_label(score, .data[[yvar]]),
      .groups = "drop"
    ) |>
    print()
}

cat(sprintf("\n── Top 10 (matrix, rule, k) combinations by mean %s ───\n", yvar))
df |>
  group_by(matrix, rule, k) |>
  summarise(
    mean_yield = round(mean(.data[[yvar]], na.rm = TRUE), 2),
    .groups = "drop"
  ) |>
  slice_max(mean_yield, n = 10) |>
  print()

cat(sprintf("\nDone. Plots saved to %s/\n", opt$outdir))
